### PR TITLE
Add Protocol documentation

### PR DESCRIPTION
**Pull request by emlynmac**
**Short description of changes**
Added high-level protocol documentation of the client-server UDP interface.

**Context: Improve documentation for contributors**

**Status of this Pull Request**
Ready for peer review

**What is missing until this pull request can be merged?**
No functional code changes; just needs reviewing for accuracy.

### CODE REVIEW

**Review by ann0see — COMMENTED** [review]
Great. Thanks. Will have a look at it but clearly @softins is the expert here.

Meta: Shouldn’t this be on the website or in a docs/ folder?

**Comment by pljones** [comment]
> Meta: Shouldn’t this be on the website or in a docs/ folder? #2283

docs/ folder now it's there, I think.  It's probably something only someone writing code is going to use, so I think it's best left in the code repo.  Ideally, we'd have the API description in some meta-language that could generate source in C++ or PHP or whatever...

**Comment by emlynmac** [comment]
> > Meta: Shouldn’t this be on the website or in a docs/ folder? #2283
> 
> docs/ folder now it's there, I think. It's probably something only someone writing code is going to use, so I think it's best left in the code repo. Ideally, we'd have the API description in some meta-language that could generate source in C++ or PHP or whatever...

I’ll move it in a moment

**Comment by ann0see** [comment]
The docs folder still needs to be created by my PR

**Comment by emlynmac** [comment]
> The docs folder still needs to be created by my PR

Ah

**Comment by ann0see** [comment]
Ok. The docs folder is now created. You can now move your file into the folder (after a rebase)

File: src/JAMULUS_PROTOCOL.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+This is defined in `CServer::OnNewConnection()`
+
+The client on a new connection will:
+* Send its channel info with a `CHANNELS_INFO (0x1900)` message
+* Request the list of connected clients with a `REQ_CONN_CLIENT_LIST (0x1000)` message
+* Set the server-side jitter buffer value with a `JITT_BUF_SIZE (0x0a00)` message
+
+This is defined in `CClient::OnNewConnection()`
+
+At the end of the session, the client calls the `CLM_DISCONNECTION(0xf203)` message, until the server stops streaming audio to it.
```
Comment (hoffie) [inline src/JAMULUS_PROTOCOL.md]:
```suggestion
At the end of the session, the client calls the `CLM_DISCONNECTION (0xf203)` message, until the server stops streaming audio to it.
```

File: docs/JAMULUS_PROTOCOL.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      <------------------------------------ ACK(CHANNEL_PAN)   
+```
+
+---
+
+Audio Packet Structure
+--
+The OPUS codec is used to compress the audio over the network and the packets are documented [here](https://datatracker.ietf.org/doc/html/rfc6716).
+
+Jamulus uses a custom OPUS encoder / decoder, giving some different frame sizes, but always uses a 48kHz sample rate. OPUS and OPUS64 codecs are the only supported options currently.
```
Comment (hoffie) [inline docs/JAMULUS_PROTOCOL.md:172]: OPUS and especially OPUS64 sound like official modes, somehow. But I think it's just the names which Jamulus gave to these encoder configurations, right? So maybe that could be clarified?

File: docs/JAMULUS_PROTOCOL.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+The packet size will vary based on:
+* Stereo vs mono
+* Packet size (64/128/256)
```
Comment (hoffie) [inline docs/JAMULUS_PROTOCOL.md:176]: Hm, not sure this is totally accurate.
I think there's three default buffer sizes in the audio logic.
And there is two packet sizes (regular and small) for network buffers, so I would have expected just two values here?

File: docs/JAMULUS_PROTOCOL.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
+
+The Jamulus Audio Protocol
+==
+
+Jamulus uses connectionless UDP packets to communicate between the client and server, and additionally for directory server registration. The `src/protocol.cpp` file contains much of the details of the packets themselves, whereas this document is intended to form a higher-level view of the protocol interactions.  
+Some of the messages need to be acknowledged, some do not.
+
+All of this information can be discovered from reading the code, but hopefully is quicker to digest when available in one location. There is a wireshark disector available too, [here](https://github.com/softins/jamulus-wireshark), if you would like to inspect the packet flow.
```
Comment (hoffie) [inline docs/JAMULUS_PROTOCOL.md:8]:
```suggestion
All of this information can be discovered from reading the code, but hopefully is quicker to digest when available in one location. There is a wireshark dissector available too, [here](https://github.com/softins/jamulus-wireshark), if you would like to inspect the packet flow.
```

File: docs/JAMULUS_PROTOCOL.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+This is defined in `CServer::OnNewConnection()`
+
+The client on a new connection will:
+* Send its channel info with a `CHANNELS_INFO (0x1900)` message
+* Request the list of connected clients with a `REQ_CONN_CLIENT_LIST (0x1000)` message
+* Set the server-side jitter buffer value with a `JITT_BUF_SIZE (0x0a00)` message
+
+This is defined in `CClient::OnNewConnection()`
+
+At the end of the session, the client calls the `CLM_DISCONNECTION(0xf203)` message, until the server stops streaming audio to it.
```
Comment (hoffie) [inline docs/JAMULUS_PROTOCOL.md:68]:
```suggestion
At the end of the session, the client calls the `CLM_DISCONNECTION (0xf203)` message, until the server stops streaming audio to it.
```

**Review by hoffie — COMMENTED** [review]
Thanks a lot for providing this documentation! This is really valuable for new contributors. :)
As I don't want my review to count here as I think @softins and @pljones are way more qualified to assert correctness, I'll just leave this as a comment for now. :)

**Comment by ann0see** [comment]
Is this ready to be merged after r 3.8.2 and a review?

**Comment by softins** [comment]
> Is this ready to be merged after r 3.8.2 and a review?

Yes, I'm planning to review it, but life is rather busy right now, and it hasn't come to the top of the pile.

File: docs/JAMULUS_PROTOCOL.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
+
+The Jamulus Audio Protocol
+==
```
Comment (ann0see) [inline docs/JAMULUS_PROTOCOL.md:3]: Just a small comment on the style: Could you please follow the style like on the website (Headings are denoted with `#` not `=`).

**Review by ann0see — APPROVED** [review]
I'm not into the protocol, but it's good to have a brief intruduction here. I'm not entirely sure, but I think Jamulus handles all packages with a specific size as audio packages and all others as control message packages?

Approving either way since the document is not critical

File: docs/JAMULUS_PROTOCOL.md (hunk 6 above)
Comment (WolfganP) [reply docs/JAMULUS_PROTOCOL.md:3]: It also may be worthy to make a reference to the wireshark dissector at https://github.com/softins/jamulus-wireshark as whoever reads this doc, most surely be interested in additional troubleshooting tools.

Comment (emlynmac) [reply docs/JAMULUS_PROTOCOL.md:3]: It's mentioned in line 8?

Comment (emlynmac) [reply docs/JAMULUS_PROTOCOL.md]: > Just a small comment on the style: Could you please follow the style like on the website (Headings are denoted with `#` not `=`).

Done.

**Review by ann0see — APPROVED** [review]
Thanks.

**Review by softins — COMMENTED** [review]
This document is a good start, and a descriptive "what happens" is definitely needed and useful.

I do have some comments, though, noted down as I read through:

- In the message packet structure, it needs to be specified that 2-byte numeric items are sent in little-endian order. So a length of 300, for example, would be a byte of 0x2C (44) followed by a byte of 0x01 (for 256).

- Related to the above point, the message types should be expressed as native numbers, e.g. `CLIENT_ID (32, 0x0020)`, `CLM_DISCONNECTION (1010, 0x03f2)`. These numbers then correspond to those in `src/protocol.h`, and those displayed in the Wireshark dissector. But in the packet, for example, the `CLM_DISCONNECTION` ID would be 0xf2 followed by 0x03.

- "Some ... need to be acknowledged, some do not." could be elaborated to say that all connected-state messages (ID < 1000) need to be acknowledged, except for ACK itself, and connectionless messages (CLM_xxx, ID >= 1000) do not need to be acknowledged.

I'm happy either to make and push these changes myself, or to leave it to @emlynmac.

**Comment by emlynmac** [comment]
> 

Thanks for the review @softins.
For the acknowledgment of IDs < 1000, is there a time limit by which this must happen or a retransmit occurs? Would be good to include that too.

I've added the decimal values to the hex in the flows.

**Comment by softins** [comment]
> Thanks for the review @softins. For the acknowledgment of IDs < 1000, is there a time limit by which this must happen or a retransmit occurs? Would be good to include that too.

Yes, the time limit is 400ms, which is defined in https://github.com/jamulussoftware/jamulus/blob/7f9f039e06a9734cec3639f575c5e7b31cbd1a77/src/protocol.h#L94-L95

> I've added the decimal values to the hex in the flows.

Thanks. I think the hex versions should have their bytes swapped in the descriptions too, so they match the decimal.

It would have been nice if the protocol had used network byte order, but we're years beyond that now, for Jamulus!

**Comment by emlynmac** [comment]
> Thanks. I think the hex versions should have their bytes swapped in the descriptions too, so they match the decimal.

I deliberately put them in this way around, so that the hex values match the Wireshark dissector trace values. It made it easier to see

**Comment by ann0see** [comment]
@softins is this now ready?

**Comment by softins** [comment]
> @softins is this now ready?

Not sure, but I have no time to review again until Friday.

File: docs/JAMULUS_PROTOCOL.md (hunk 2 above)
Comment (softins) [reply docs/JAMULUS_PROTOCOL.md:166]: Yes, I've just looked through the code, and the only difference between them is that `OPUS` configures the Opus codec to use 128 samples per frame, and `OPUS64` configures it to use 64 samples per frame. But it's the same codec, just two different Jamulus modes.

File: docs/JAMULUS_PROTOCOL.md (hunk 3 above)
Comment (softins) [reply docs/JAMULUS_PROTOCOL.md:170]: I think this line should refer to buffer size:
```suggestion
* Buffer size (64/128/256 samples)
```

**Review by softins — APPROVED** [review]
Subject to a couple of comments, I'm happy to approve it. It would be good as time permits to expand it, but I don't think there's a reason to keep it languishing as a PR.

**Comment by ann0see** [comment]
CHANGELOG: Docs: Add technical documentation for Jamulus protocol

**Pull request merged**